lib/shims/eq/R2015b/isfolder.m

```diff
@@ -0,0 +1,9 @@
+function result = isfolder(folderName)
+
+% ISFOLDER Introduced in R2017b
+% ISFOLDER Determine if input is folder
+% https://www.mathworks.com/help/matlab/ref/isfolder.html
+
+result = isdir(folderName);
```

Comment:
`isdir` is not actually equivalent to `isfolder`, because `isdir` searches the Matlab path when you give it a non-qualified name, so it will return different results and false-positives in some cases. Need to do a different implementation that gives the same behavior as `isfolder`, which is a strict check for existence and is-a-directory for the given input path.

If there's not an easy way to do this with native Matlab functions, you can call out to Java to do it:

```matlab
jFile = java.io.File(folderName);
out = jFile.isDirectory();
```

I think that has the exact same semantics as `isfolder`.

Reply:
I take your point that isfolder() is different from isdir(), and the proposed 2015b shim is not a perfect reimplementation of isfolder().  But for SLF4M, the only use of isfolder() is for a fully qualified path, so I'm thinking the simplistic re-implementation is sufficient as a shim for SLF4M.
     
```
     26       thisFile = mfilename('fullpath');
     27       distDir = fileparts(fileparts(fileparts(fileparts(thisFile))));
     28       libDir = fullfile(distDir, 'lib');
     29
     30       % Load Dispstr
     31       matlabLibDir = fullfile(libDir, 'matlab');
     32       dispstrMcodeDir = fullfile(matlabLibDir, 'dispstr', 'dispstr-1.1.1', 'Mcode');
     33       mPath = strsplit(path, pathsep);
     34       if ~ismember(dispstrMcodeDir, mPath)
     35         addpath(dispstrMcodeDir);
     36       end
     37
     38       % Compatibility shims
     39       % TODO: Replace "eq" logic with lt/ge logic
     40
     41       matlabRel = ['R' version('-release')];
     42       shimsDir = fullfile(libDir, 'shims', 'eq', matlabRel);
     43       if isfolder(shimsDir)
```

Reply:
But once you load in SLF4M with the compatibility shims, `isfolder` is on the path and can be invoked by any code, not just the code inside SLF4M. Matlab doesn't have library modularity yet. So any shimmed functions should be maximally compatible/complete.

Alternately, all SLF4M's calls to `isfolder` could be replaced with calls to a different `+internal` or `private` function, so SLF4M no longer has a dependency on `isfolder` and doesn't need to shim it.

Reply:
The assumption is that if I'm shimming for R2015b, then the rest of the code is strictly R2015b, and there are no other `isfolder `calls.
But yes, if the R2015b code has other `isfolder `calls, then this shimming causes a problem :-/.  
My objective for the 4 shims, given the assumption above, was minimally compatible/complete, not maximally.  
If I attempt maximally, then I have to improve `contains.m` also, and also `@char/string.m` seems like a real challenge.
This seems like a fundamental issue with the shims approach.

Your alternative -- replace all calls to `isfolder`, `newline`, `string`, & `contains` with calls to a +internal or private `myisfolder`, `mynewline`, `mystring`, & `mycontains`, which in turn do the right things, is certainly cleaner.

Or there could be a library installation step, where the private minimally compatible implementations are copied into place only if that library installation step had been performed.
The library initialization call would ensure that the matlab version and the localization (if any) is matched.  So the private localizations would include a simple iscompatible as well.
I'm not pitching this a great idea, I'm just brainstorming out loud.

Reply:
So, the particular scenario I have in mind for compatibility reasons is when a user is using SLF4M in R2015b-R2017a in combination with other code which _itself_ has shimmed these functions by providing their own implementations that are forward-compatible with the later-introduced Matlab functions. (Later Matlab's behavior is the only real reference we have here; if they implemented functions with those names but different behavior, then all bets are off and there's nothing we can do. If SLF4M also provides fully-future-Matlab-compatible shim versions of those functions, then if SLF4M gets on the path ahead of their user code, their code will still work, and vice versa.

> If I attempt maximally, then I have to improve contains.m also, and also @char/string.m seems like a real challenge.

Oh yeah. `contains` would be hard, and a full `string` implementation would probably be larger than the SLF4M library itself.

So, what if a user wrote some R2015b code, and their application has a notion of a "folder" format that is actually a ZIP file with some documents inside it, and their code has an `isfolder` function that tests for that? That would be a totally legit thing to do; in R2015b you had no way of knowing that "isfolder" was a name that Matlab would user later! Or they wrote a GUI app with some tree components, and to them a "folder" is an internal tree node or some other GUI object?

An installation step is a no-go I think; that's hard to do in Matlab, especially in the scenario I have for SLF4M where SLF4M is incorporated as a dependency in other developers' Matlab code which itself is then redistributed. And it's inconvenient for developers of the code.

This conversation is making me reconsider the wisdom of this whole shimming approach, and think maybe `+internal` is a better idea, and worth the effort of switching to if we want SLF4M to support older Matlab versions like this. Sorry to make a U-turn and request your whole PR be done differently! Let's mull this over for a little bit.

Reply:
Okay, I've thought about this, and... now I think that "shimming" for forward-compatibility is actually a bad idea for something like SLF4M which is intended to be reused as a library inside other Matlab "applications" or programs.

Part of the API of a particular version of Matlab is its "negative space" - the set of identifiers which Matlab does not define semantics for, and are thus up to its users to do with what they will.

For a Matlab "library" - a piece of code which is intended to be combined with other code, like SLF4M is - if it wants to support earlier versions of Matlab, it should respect that negative space by not "grabbing" identifiers in it and imposing semantics on thm.

So. I now think that, if SLF4M wants to support earlier versions of Matlab (which it does, because it has users like you, who cared enough to report this issue!), it should avoid shimming, and instead use internal implementations which work in either Matlab version, so shimming is unnecessary.

So instead of providing R2015b shims, let's rip out the whole shimming mechanism and use an internal implementation that doesn't depend on these Matlab-supplied functions at all for this behavior.

I'm sorry to do a U-turn on you here and say "sorry, that's not how I think we should do things" after you've already submitted a PR. If you want to re-do this PR in a non-shimming manner, go for it. But I'll try to find some time myself (maybe over this weekend?) to re-do the SLF4M internals to remove the dependencies on `isfolder`, `contains`, and newer functions, so that it can work on R2015b etc. in a simpler manner and shimming is just not required.

Reply:
No need for concern; I love this discussion.  I'm learning things, and I feel like I'm contributing more good than bad, net positive :).

Does it make sense to accept only the 2-line bugfix to Log4jConfigurator.m, and then consider this PR closed?  Shims in this PR are simply rejected.  (I'm a git novice -- I'm assuming you can select parts of a PR to accept instead of all-or-nothing).  As you suggested at the beginning, the R2015b compatibility belongs in a separate PR, and is an enhancement not a bugfix.

I might try again with another PR for chgs that work for R2015b (as far as I can tell).  If I do, and it's redundant to your work-in-progress and ultimately not acceptable, no worries.

Reply:
> I love this discussion.

Dude me too. Most of the whole reason I got in to FLOSS Matlab development is to meet other Matlab developers and have conversations like this.

And this is useful for me too! I think this current "shimming is bad for a *library* that wants to support a broad range of Matlab versions" view is one that is good, and would have taken me a lot longer to arrive at myself if we hadn't had this discussion.

> Does it make sense to accept only the 2-line bugfix to Log4jConfigurator.m, and then consider this PR closed?

Yes.

> I'm assuming you can select parts of a PR to accept instead of all-or-nothing

Nope! PRs are in fact all-or-nothing. I can't modify the code in your PR myself (kinda), because that code is in a branch in your repo and has your author name attached to it. So if you want to modify this PR (by revising and force-pushing the code on your branch) to remove the shimming changes and just have the Log4jConfigurator changes, I will merge it.

And then I think I will go through the SLF4M code myself and tear out all the shimming stuff and replace it with more encapsulated code that is "safer" and compatible with more Matlab versions.

> I might try again with another PR for chgs that work for R2015b (as far as I can tell).

Nah, don't bother. Now that I know there's an actual user out there who wants R2015b compatibility in SLF4M, I'm happy to do this myself. And I want to play around with the code and figure out what a good replacement for the shimming stuff is, while keeping the code "tidy". And that's very much dependent on my personal taste in code design, so it'll be easier for me to just do myself.

Reply:
simplest path fwd is you reject the PR, and you do your own commit for the 2-line bugfix to Log4jConfigurator.m.  Right?  I'll sniff around; maybe I can even withdraw the PR from my side.